ibis/common/tests/test_deferred.py

```diff
@@ -0,0 +1,540 @@
+
+    b = Attr(MyType(1, 2), "a")
+    assert b.build({}) == 1
+
+
+def test_builder_item():
+    v = Variable("v")
+    b = Item(v, 1)
+    assert b.build({"v": [1, 2, 3]}) == 2
+
+    b = Item(FrozenDict(a=1, b=2), "a")
+    assert b.build({}) == 1
+
+
+def test_builder_mapping():
+    b = Mapping({"a": 1, "b": 2})
+    assert b.build({}) == {"a": 1, "b": 2}
+
+    b = Mapping({"a": Just(1), "b": Just(2)})
+    assert b.build({}) == {"a": 1, "b": 2}
+
+    b = Mapping({"a": Just(1), "b": Just(2), "c": _})
+    assert b.build({"_": 3}) == {"a": 1, "b": 2, "c": 3}
+
+
+def test_builder():
+    def fn(x, ctx):
+        return x + 1
+
+    assert builder(1) == Just(1)
+    assert builder(Just(1)) == Just(1)
+    assert builder(Just(Just(1))) == Just(1)
+    assert builder(fn) == Factory(fn)
+    assert builder(()) == Sequence(())
+    assert builder((1, 2, _)) == Sequence((Just(1), Just(2), _))
+    assert builder({}) == Mapping({})
+    assert builder({"a": 1, "b": _}) == Mapping({"a": Just(1), "b": _})
+
+    assert builder(var("x")) == Variable("x")
+    assert builder(Variable("x")) == Variable("x")
+
+
+def test_builder_objects_are_hashable():
+    a = Variable("a")
+    b = Attr(a, "b")
+    c = Item(a, 1)
+    d = Call(operator.add, a, 1)
+
+    set_ = {a, b, c, d}
+    assert len(set_) == 4
+
+    for obj in [a, b, c, d]:
+        assert obj == obj
+        assert hash(obj) == hash(obj)
+        set_.add(obj)
+        assert len(set_) == 4
+
+
+@pytest.mark.parametrize(
+    ("value", "expected"),
+    [
+        ((), ()),
+        ([], []),
+        ({}, {}),
+        ((1, 2, 3), (1, 2, 3)),
+        ([1, 2, 3], [1, 2, 3]),
+        ({"a": 1, "b": 2}, {"a": 1, "b": 2}),
+        (FrozenDict({"a": 1, "b": 2}), FrozenDict({"a": 1, "b": 2})),
+    ],
+)
+def test_deferred_builds(value, expected):
+    assert deferred(value).__builder__.build({}) == expected
+
+
+def test_deferred_supports_string_arguments():
+    # deferred() is applied on all arguments of Call() except the first one and
+    # sequences are transparently handled, the check far sequences was incorrect
+    # for strings causing infinite recursion
+    b = deferred("3.14")
+    assert b.resolve() == "3.14"
+
+
+def test_deferred_object_are_not_hashable():
+    # since __eq__ is overloaded, Deferred objects are not hashable
+    with pytest.raises(TypeError, match="unhashable type"):
+        hash(_.a)
+
+
+def test_deferred_const():
+    obj = const({"a": 1, "b": 2, "c": "gamma"})
+
+    deferred = obj["c"].upper()
+    assert deferred.__builder__ == Call(Attr(Item(obj, "c"), "upper"))
+    assert deferred.resolve() == "GAMMA"
+
+
+def test_deferred_variable_getattr():
+    v = var("v")
+    p = v.copy
+    assert p.__builder__ == Attr(v, "copy")
+    assert p.__builder__.build({"v": [1, 2, 3]})() == [1, 2, 3]
+
+    p = v.copy()
+    assert p.__builder__ == Call(Attr(v, "copy"))
+    assert p.__builder__.build({"v": [1, 2, 3]}) == [1, 2, 3]
+
+
+class Table(dict):
```

Comment:
Why reimplement a table/column set of objects here rather than reusing our existing implementation? Is it just to avoid importing `ibis.expr` in tests for `ibis.common`? If so, I'd argue that's not worth the extra code in the tests file.

Reply:
Yes, that is the main reason. I prefer mocking to keep the testable logic minimal.

Reply:
Can you add a comment above the mocked api saying so (and maybe use the word "Mock" in the class names)? I had to read a bit to understand that `Column` wasn't an ibis column but a local class definition.

Reply:
Sure, going to rename them.

Reply:
Renamed.